src/super_context.ml

```diff
@@ -237,10 +247,52 @@ let ocaml_flags t ~dir (x : Buildable.t) =
     ~default:(Env.ocaml_flags t ~dir)
     ~eval:(Expander.expand_and_eval_set expander)
 
+let c_flags t ~dir ~(lib : Library.t) ccg =
+  let expander = Env.expander t ~dir in
+  let eval = Expander.expand_and_eval_set expander in
+  let flags = lib.c_flags in
+  let default = Env.c_flags t ~dir in
+  Build.memoize "c flags"
+    begin
+    if Ordered_set_lang.Unexpanded.has_special_forms flags then
+      let c = eval flags ~standard:default in
+      let open Build.O in (c >>^ fun l -> l @ ccg)
+    else
+      eval flags ~standard:(Build.return ccg)
+  end
+
+
+let cxx_flags_gather t ~dir ~(lib : Library.t) ccg =
```

Comment:
Why do we need the `_gather` suffix here?

Reply:
There is another cxx_flags function in Super_context, and a corresponding field of t.  The existing function inspects context.ocamlc_flags and from them computes cxx_flags. Would it be better to rename existing cxx_flags to something like cxx_flags_from_ocamlc_flags or cxx_flags_default?

Reply:
I think that it would be better to remove the original function. It seems like we shouldn't really be using it anymore as the `cxx_flags` are now per directory.

Reply:
The original cxx_flags has been used in combination with expanded lib.cxx_flags in lib_rules.ml to build_cxx_file. The new cxx_flags_gather replaces lib.cxx_flags. I am not sure why it makes the original cxx_flags obsolete. For that matter, I don't know why the original cxx_flags were computed in this way from ocamlc_flags and what guarantees that C++ compiler knows about all possible ocamlc_flags (except -std) used in a project. The must be some assumptions about the use of these cxx_flags that I am missing.

In any case, I attempted to make this change in two steps:
1) change Super_context to rename the original cxx_flags to cxx_flags_orig and renamed cxx_flags_gather to cxx_flags.
55a4d224be5210ffeba7c551b4ad7bb2bd1a745b
2) remove cxx_flags_orig from Super_context.  
They are no longer used anywhere except in Pform.create, where they are getting an akward special treatment. I replaced it by recomputing them (as originally in super_context) and it is analogous to the way cflags are obtained in Pform.
61045a521d6646b5a44780e2fab3c39cd71f8c75

My main question is whether cxx_flags_orig need to be conjoined to the new cxx_flags or simply dropped from build_cxx_files in lib_rule. For now, I dropped them, but it might not be safe: removing the original cxx_flags might break other things in dune or external projects that rely on these flags being inferred from ocamlc_flags.

There is a test in the testsuite for it that passes. I haven't had a chance to test it in a project that actually compiles, because the stubs in my project are in C, but I can create such a test, if this solution seems good to you.

Reply:
Thank you for the explanations! It makes sense. I try to add a test for (4).

Reply:
> > My main question is whether cxx_flags_orig need to be conjoined to the new cxx_flags or simply dropped from build_cxx_files in lib_rule. For now, I dropped them, but it might not be safe: removing the original cxx_flags might break other things in dune or external projects that rely on these flags being inferred from ocamlc_flags.
> 
> Those are the defaults so they should be inherited. If a user wants to drop them, then setting them explicitly is always possible (without the use of `:standard`).

It looks like  cxx_flags_orig was always added in lib_rules, regardless of ":standard". 
After removing this default in commit 61045a521d6646b5a44780e2fab3c39cd71f8c75, I have noticed that one package from opam fails to build with this version.
The package is re2 and the problem is in 
https://github.com/janestreet/re2/blob/master/src/jbuild.
Adding -fPIC at least on my machine fixes the problem, not sure if that's a good solution in general.

Reply:
> It looks like cxx_flags_orig was always added in lib_rules, regardless of ":standard".
> After removing this default in commit [61045a5](https://github.com/ocaml/dune/commit/61045a521d6646b5a44780e2fab3c39cd71f8c75), I have noticed that one package from opam fails to build with this version.
> The package is re2 and the problem is in
> https://github.com/janestreet/re2/blob/master/src/jbuild.
> Adding -fPIC at least on my machine fixes the problem, not sure if that's a good solution in general.

This is now fixed. Thanks to @xclerc for help!